mesonbuild/build.py

```diff
@@ -831,6 +834,11 @@ def __init__(
         self.link(link_targets)
         self.link_whole(link_whole_targets)
 
+        if self.uses_vala():
```

Comment:
One alternative is to add a `def has_vala_vapi(self) -> bool` method, and override it in Executable. This way you never have stuff that is set and then reset.

Otherwise looks good.

Reply:
I think that was a worthwhile thing to have lunch first. I've used a protected method in this version to set the attributes, so that Executable can override that method to only set them to not None in the export dynamic case.